Privacy whitelist and resolving list handling

carlescufi opened the pull request.

carlescufi committed:
Bluetooth: controller: Add whitelist population with privacy

When using privacy, an additional cache of the actual privacy peers is
required to avoid additional processing in the ISR (since some of the
peers in the whitelist will be disabled by the corresponding privacy
settings). Add the cache and populate the actual whitelist just before
advertising, scanning or initiating a connection.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
Bluetooth: controller: Properly remove peer IRKs from RL

When removing a device from the resolving list, if it contained a peer
IRK then it is necessary to update the indices that point to the peer
IRK list, since the list itself is contiguous in memory.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
Bluetooth: controller: Add RL filter

In order for privacy to work correctly with and without peer IRKs, an
additional hardware filter is required to help match the peer device in
the case address resolution is not possible for a peer since the Host
has not provided a peer IRK for it.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
Bluetooth: controller: Handle Privacy Modes

Add handling of the LE Set Privacy Mode HCI Command along with the logic
required to take it into account when populating the hardware filters.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ll_filter.c:
```diff
@@ -64,29 +142,37 @@ static void filter_clear(struct ll_filter *filter)
 static u32_t filter_add(struct ll_filter *filter, u8_t addr_type, u8_t *bdaddr)
 {
-	u8_t index;
+	int index;
```
Why int? parameter to filter_insert is u8_t.

carlescufi replied:
I will change the parameter, I prefer to use ints whenever no specific width is required

cvinayak replied:
ok. agreed.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ll_filter.c:
```diff
@@ -64,29 +142,37 @@ static void filter_clear(struct ll_filter *filter)
 			   u8_t *bdaddr)
 {
-	u8_t index;
+	int index;
```
Why int?

carlescufi replied:
Because I don't care about the width

cvinayak replied:
ok. agreed.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ll_filter.c:
```diff
@@ -157,21 +251,77 @@ u32_t ll_wl_remove(bt_addr_le_t *addr)
+static void filter_wl_update(void)
+{
+	int i;
+	/* Populate filter from wl peers */
```
Missing blank line

cvinayak commented on subsys/bluetooth/controller/ll_sw/ll_filter.c:
```diff
@@ -51,11 +64,76 @@ static struct rl_dev {
+{
+	/* find the device and mark it as empty */
+	int i = wl_peers_find(id_addr->type, id_addr->a.val);
+	if (i >= 0) {
```
Missing blank line.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ll_filter.c:
```diff
@@ -157,21 +251,77 @@ u32_t ll_wl_remove(bt_addr_le_t *addr)
 
+	for (i = 0; i < WL_SIZE; i++) {
+		int j = wl_peers[i].rl_idx;
+		if (!rl_enable || j == IDX_NONE || !rl[j].pirk || rl[j].dev) {
```
Missing blank line.

cvinayak replied:
rl_enable should not be checked inside loop.

carlescufi replied:
rl_enable is only modified in thread mode by cooperative threads

cvinayak replied:
if you have the check for rl_enable outside the loop, you just need to loop i times calling filter_insert without carrying for reading j. This is just a optimization, more code, but faster by few cycles for each address.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ll_filter.c:
```diff
@@ -157,21 +251,77 @@ u32_t ll_wl_remove(bt_addr_le_t *addr)
+static void filter_rl_update(void)
 {
 	int i;
+	/* No whitelist: populate filter from rl peers */
```
Missing blank line

cvinayak commented on subsys/bluetooth/controller/ll_sw/ll_filter.c:
```diff
@@ -157,21 +251,77 @@ u32_t ll_wl_remove(bt_addr_le_t *addr)
+}
+
+
+void ll_filters_update(int adv_fp, int scan_fp)
```
Have two different functions, saves on param passing and program stack usage.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ll_filter.c:
```diff
@@ -157,21 +251,77 @@ u32_t ll_wl_remove(bt_addr_le_t *addr)
+int ll_rl_find(u8_t id_addr_type, u8_t *id_addr)
+{
+	int i;
+	for (i = 0; i < CONFIG_BLUETOOTH_CONTROLLER_RL_SIZE; i++) {
```
blank line

cvinayak commented on subsys/bluetooth/controller/ll_sw/ll_filter.c:
```diff
@@ -384,12 +534,18 @@ u32_t ll_rl_add(bt_addr_le_t *id_addr, const u8_t pirk[16],
 		return BT_HCI_ERR_CMD_DISALLOWED;
 	}
 
+	i = ll_rl_find(id_addr->type, id_addr->a.val);
```
lets have ll_rl_find(type, addr, *free_idx); so that if not found use free_idx to insert device. Saves a subsequest loop to find empty slot.

carlescufi commented on subsys/bluetooth/controller/ll_sw/ll_adv.c:
```diff
@@ -384,14 +384,17 @@ u32_t ll_adv_enable(u8_t enable)
 		bool priv = false;
 #if defined(CONFIG_BLUETOOTH_CONTROLLER_PRIVACY)
+		/* Prepare whitelist and optionally resolving list */
+		ll_filters_update(ll_adv.filter_policy, -1);
```
no, we are updating 2 filters

carlescufi commented on subsys/bluetooth/controller/ll_sw/ll_filter.h:
```diff
@@ -13,11 +13,12 @@ struct ll_filter {
 };
 
 void ll_filter_reset(bool init);
+void ll_filters_update(int adv_fp, int scan_fp);
```
Well, logically this fits better together: a single function to update the filters whenever activity starts. I don't mind splitting this up though, I'll take a look

carlescufi commented on subsys/bluetooth/controller/ll_sw/ll_scan.c:
```diff
@@ -25,7 +26,7 @@ static struct {
 
 	u8_t  tx_addr:1;
-	u8_t  filter_policy:1;
+	u8_t  filter_policy:2;
```
Yes, they are 2 bits in the structures and they use u8_t as parameters

cvinayak replied:
ok.

carlescufi committed:
Bluetooth: controller: Optimize RL addition

To avoid iterating twice through the list, have ll_rl_find() return the
first free empty slot on the list.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>
